### PR TITLE
Add set of modular Orka tasks

### PR DESCRIPTION
**Pull request by spikeburton**
# Changes

This change adds a set of modular Tasks to integrate Tekton with Orka
by MacStadium. Orka allows users to run macOS builds and workloads on
real Apple hardware. An Orka / Tekton integration will allow the
utilization of macOS build agents for CI jobs that run on Tekton
pipelines.

The tasks orka-init, orka-deploy and orka-teardown are designed to work
together in order to allow the utilization of multiple macOS
build agents in a CI/CD pipeline. The orka-deploy Task allows the user
to provide a build script as a parameter, and will allow the user to
store build artifacts in a Tekton workspace. The orka-init and
orka-teardown tasks create and delete virtual machine configurations.
This is accomplished by making calls to the Orka API running in a
separate Kubernetes cluster over a VPN connection.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality of task changed or new task added)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Complies with [Catalog Organization TEP][TEP], see [example]. **Note** [An issue has been filed to automate this validation][validation]
  - [x] File path follows  `<kind>/<name>/<version>/name.yaml`
  - [x] Has `README.md` at `<kind>/<name>/<version>/README.md`
  - [x] Has mandatory `metadata.labels` - `app.kubernetes.io/version` the same as the `<version>` of the resource
  - [x] Has mandatory `metadata.annotations` `tekton.dev/pipelines.minVersion`
  - [x] mandatory `spec.description` follows the convention

          ```

          spec:
            description: >-
              one line summary of the resource

              Paragraph(s) to describe the resource.
          ```

_See [the contribution guide](https://github.com/tektoncd/catalog/blob/master/CONTRIBUTING.md)
for more details._

---

[TEP]: https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md
[example]: https://github.com/tektoncd/catalog/tree/master/task/git-clone/0.1
[validation]:  https://github.com/tektoncd/catalog/issues/413

### CODE REVIEW

**Comment by spikeburton** [comment]
/assign @dlorenc

I am not sure why the integration tests are failing. Branch is current with master and tests pass when run locally. Any insight here?

**Comment by chmouel** [comment]
The CI fail because according to the catalog spec (for discoverability on tekton hub) it's one task in one directory, you have here multiple of them in orka-deploy

**Comment by spikeburton** [comment]
> The CI fail because according to the catalog spec (for discoverability on tekton hub) it's one task in one directory, you have here multiple of them in orka-deploy

@chmouel I have one Task per directory - please view file tree here:
https://github.com/macstadium/catalog/tree/orka-modular-0.1/task/orka-deploy/0.1

**Comment by chmouel** [comment]
I see, the test runner gets confused by the mockhttps://github.com/macstadium/catalog/tree/orka-modular-0.1/task/orka-deploy/0.1/tests/mocks/orka.yaml

either we fix the test runner or have this directory moved somewhere else if that's possible..

**Comment by spikeburton** [comment]
> I see, the test runner when collecting which test to tests for this commmit [1] gets confused by the mock https://github.com/macstadium/catalog/tree/orka-modular-0.1/task/orka-deploy/0.1/tests/mocks/orka.yaml
> 
> either we fix the test runner or have this directory moved somewhere else if that's possible..
> 
> I think @vinamra28 did a fix for this last week or maybe that brokes it ? (in the way we detect change in PR we should avoid subdir)
> 
> [1] which is why you are not seeing this issue when running directly with run-tests.sh

@chmouel @vinamra28 when could a fix for the test runner be expected? What is the alternative approach for how the file tree should be arranged?

**Comment by vinamra28** [comment]
> I see, the test runner when collecting which test to tests for this commmit [1] gets confused by the mock https://github.com/macstadium/catalog/tree/orka-modular-0.1/task/orka-deploy/0.1/tests/mocks/orka.yaml
> 
> either we fix the test runner or have this directory moved somewhere else if that's possible..
> 
> I think @vinamra28 did a fix for this last week or maybe that brokes it ? (in the way we detect change in PR we should avoid subdir)

the fix that I did last week was to detect any file change in the test dir but yeah that seems to broke the CI :disappointed:. It's detecting subdirs as well. Will try to fix this asap

**Comment by spikeburton** [comment]
Looks like the service account is unable to create / delete resources (secrets / config maps) in the cluster used by the test runner ... any suggestions?

**Comment by spikeburton** [comment]
> Looks like the service account is unable to create / delete resources (secrets / config maps) in the cluster used by the test runner ... any suggestions?

@chmouel @vinamra28 any insight here?

**Comment by vinamra28** [comment]
@spikeburton the tasks are running in parallel so maybe there is some clashing between the names of `ClusterRole` and `ClusterRoleBinding` of each task. Can you please try renaming them for all three tasks? I am hoping that this should work fine :crossed_fingers:

**Comment by spikeburton** [comment]
@vinamra28 that did the trick - thank you ;)

**Review by vinamra28 — COMMENTED** [review]
not a blocker just a thought

File: task/orka-deploy/0.1/orka-deploy.yaml [hunk 1]
```diff
@@ -0,0 +1,150 @@
+---
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: orka-deploy
+  labels:
+    app.kubernetes.io/version: "0.1"
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.16.0"
```
[the remark this answers is not in the document]
Comment (spikeburton) [reply task/orka-deploy/0.1/orka-deploy.yaml:9]: Good idea 👍

File: task/orka-teardown/0.1/tests/pre-apply-task-hook.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+${KUBECTL_CMD} delete clusterrole orka-runner --ignore-not-found
+${KUBECTL_CMD} delete clusterrolebinding orka-runner --ignore-not-found
```
Comment (vinamra28) [inline task/orka-teardown/0.1/tests/pre-apply-task-hook.sh:4]: do we need this now? :thinking:  or probably we can rename these also

File: task/orka-init/0.1/tests/pre-apply-task-hook.sh [hunk 3]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+${KUBECTL_CMD} delete clusterrole orka-runner --ignore-not-found
+${KUBECTL_CMD} delete clusterrolebinding orka-runner --ignore-not-found
```
Comment (vinamra28) [inline task/orka-init/0.1/tests/pre-apply-task-hook.sh:4]: same can we rename these as well or remove if not required

File: task/orka-init/0.1/orka-init.yaml [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+      description: |
+        The name of the config map, which stores the name of the generated VM configuration.
+      default: orka-vm-name
+    - name: orka-vm-name-config-key
+      type: string
+      description: |
+        The name of the key in the VM name config map, which stores the name of the generated VM configuration.
+      default: vm-name
+  results:
+    - name: vm-name
```
Comment (vinamra28) [inline task/orka-init/0.1/orka-init.yaml:73]: are we storing any results?

File: task/orka-deploy/0.1/tests/pre-apply-task-hook.sh [hunk 5]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+${KUBECTL_CMD} delete clusterrole orka-runner --ignore-not-found
+${KUBECTL_CMD} delete clusterrolebinding orka-runner --ignore-not-found
```
Comment (vinamra28) [inline task/orka-deploy/0.1/tests/pre-apply-task-hook.sh:4]: same can we rename or remove these?

File: task/orka-init/0.1/orka-init.yaml (hunk 4 above)
Comment (spikeburton) [reply task/orka-init/0.1/orka-init.yaml:73]: Good catch - this is left over from a previous implementation and is no longer used.

File: task/orka-teardown/0.1/tests/pre-apply-task-hook.sh (hunk 2 above)
Comment (spikeburton) [reply task/orka-teardown/0.1/tests/pre-apply-task-hook.sh]: No longer needed - removed.

**Review by vinamra28 — COMMENTED** [review]
/lgtm
Thank you for the PR :+1: 
/cc @chmouel @sbwsg @vdemeester

**Comment by chmouel** [comment]
Looking good, I don't think you need to have the ClusterRoles in test, you can just have a local Role, it doesn't matter much as long we make sure they are not conflicting, but out of curiosity, why did you choose `ClusterRoles`

/lgtm

**Comment by spikeburton** [comment]
> Looking good, I don't think you need to have the ClusterRoles in test, you can just have a local Role, it doesn't matter much as long we make sure they are not conflicting, but out of curiosity, why did you choose `ClusterRoles`
> 
> /lgtm

Good point, I just kept it consistent for the tests as how the Tasks function. The idea is that you are not limited to running your workloads in a single namespace. For the tests, the roles are all uniquely named, so there should be no issues.

**Comment by chmouel** [comment]
thanks!

/approve

**Pull request merged**